packages/runtimes/react-refresh/src/ReactRefreshRuntime.js

```diff
@@ -0,0 +1,40 @@
+// @flow strict-local
+
+import {Runtime} from '@parcel/plugin';
+import path from 'path';
+import invariant from 'assert';
+
+export default new Runtime({
+  apply({bundle, options}) {
+    if (bundle.type !== 'js' || !bundle.env.isBrowser() || !options.hot) {
+      return;
+    }
+
+    let mainEntry = bundle.getMainEntry();
+    let code;
+    if (mainEntry && mainEntry.meta.reactRefreshRuntimePath !== undefined) {
+      let runtime = mainEntry.meta.reactRefreshRuntimePath;
+      invariant(typeof runtime === 'string');
+
+      code = `
+var Refresh = require('${path.relative(__dirname, runtime)}');
```

Comment:
Could you just put `require('react-refresh/runtime')` here? Then, below set `filePath` to be `path.join(path.dirname(mainEntry), 'react-refresh-entry.js')`. That way, it will resolve the `react-refresh` module relative to the same directory as the main entry. That would avoid the need to resolve it in the transformer, and also the hack to check if it's in the monorepo (path would contain node_modules, so `isSource` would be correct).

Reply:
Then react-refresh has to be a dependency of the runtime as well and the react-refresh instance in the transformer and in the runtime **has** to be the same

Reply:
oh, well if it's a dep of both `@parcel/transformer-react-refresh-wrap` and `@parcel/runtime-react-refresh`, then you could set `filePath` to `__filename` like you have it and it would resolve relative to the runtime.

Reply:
> well if it's a dep of both

that doesn't work, it wouldn't guarantee that both use the same `react-refresh` asset/module in the browser. Just like React doesn't really like it if there are two instances of it running on one page.

Reply:
> Would only be different if we mess it up

We would rely on the package manager that is installing parcel to hoist properly.

Reply:
oh - are there npm-alikes that don't do that?

Reply:
Theoretically, the version ranges could be resolved to different versions (not sure if this could happen when updating), you could craft a lockfile that forces different versions for the two packages (unless we pin the version).
Not sure if a package manager still does that. But I think this is an implementation detail that was never part of the node_modes "spec".

Reply:
I have made them peerDependencies (and only one plugin with a "hard" dependency), now I'd say it isn't our fault 🤷‍♂